rascal-lsp/src/main/java/org/rascalmpl/vscode/lsp/util/SemanticTokenizer.java

```diff
@@ -102,35 +102,73 @@ public SemanticTokensCapabilities capabilities() {
 
     private static class TokenList {
         private final List<Integer> theList = new ArrayList<>(500);
-        private int previousLine = 0;
-        private int previousStart = 0;
+        private int previousLineAbsolute = 0;
```

Comment:
as the statemachine around previous & current is getting bigger, I'm starting to wonder if we should do something where we collect the "current" token, and only flush when the category changes (or there is a gap). Such that we only write to `theList` once we've found a new token (or are at the end of the tree).

Reply:
Yes, that would be a more direct way of doing it. I haven't investigated in detail, but the logic around the usage of `TokenList` (particularly in the `collect...` methods) makes this a bit hard to do. I think we should do improve this as part of the nested `syntax` non-terminal thing (which probably requires changes to the logic in the `collect...` methods anyway).